From 26dd02eefd36c2a47dd5a7e419e5b3b9d7a718b3 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 27 Feb 2024 15:04:35 +0800
Subject: [PATCH 068/236] mtk: hostapd: specify link id for unicast DEAUTH

When deauthenticating the STA, hostapd should specifies the setup link
of the target STA so that the TX status of the DEAUTH can be forwarded
to the correct link (BSS).

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ap_drv_ops.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index e1d71218c..f11497134 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -931,7 +931,11 @@ int hostapd_drv_sta_deauth(struct hostapd_data *hapd,
 	if (hapd->conf->mld_ap) {
 		struct sta_info *sta = ap_get_sta(hapd, addr);
 
-		link_id = hapd->mld_link_id;
+		if (sta)
+			link_id = sta->mld_assoc_link_id;
+		else
+			link_id = hapd->mld_link_id;
+
 		if (ap_sta_is_mld(hapd, sta))
 			own_addr = hapd->mld->mld_addr;
 	}
-- 
2.45.2

